src/core/net/srp_client.cpp

```diff
@@ -0,0 +1,1559 @@
+{
+    HostInfo            hostInfoCopy = mHostInfo;
+    LinkedList<Service> removedServices;
+
+    if (mHostInfo.GetState() == kRemoved)
+    {
+        mHostInfo.Clear();
+    }
+
+    ResetRetryWaitInterval();
+    SetState(kStateUpdated);
+
+    GetRemovedServices(removedServices);
+    InvokeCallback(OT_ERROR_NONE, hostInfoCopy, removedServices.GetHead());
+}
+
+void Client::GetRemovedServices(LinkedList<Service> &aRemovedServices)
+{
+    Service *service;
+    Service *prev;
+    Service *next;
+
+    for (prev = nullptr, service = mServices.GetHead(); service != nullptr; service = next)
+    {
+        next = service->GetNext();
+
+        if (service->GetState() == kRemoved)
+        {
+            mServices.PopAfter(prev);
+            aRemovedServices.Push(*service);
+
+            // When the service is removed from the list
+            // we keep the `prev` pointer same as before.
+        }
+        else
+        {
+            prev = service;
+        }
+    }
+}
+
+otError Client::ReadResourceRecord(const Message &aMessage, uint16_t &aOffset, Dns::ResourceRecord &aRecord)
+{
+    // Reads and skips over a Resource Record (RR) from message at
+    // given offset. On success, `aOffset` is updated to point to end
+    // of RR.
+
+    otError error;
+
+    SuccessOrExit(error = Dns::Name::ParseName(aMessage, aOffset));
+    SuccessOrExit(error = aMessage.Read(aOffset, aRecord));
+    VerifyOrExit(aOffset + aRecord.GetSize() <= aMessage.GetLength(), error = OT_ERROR_PARSE);
+    aOffset += static_cast<uint16_t>(aRecord.GetSize());
+
+exit:
+    return error;
+}
+
+otError Client::ProcessOptRecord(const Message &aMessage, uint16_t aOffset, const Dns::OptRecord &aOptRecord)
+{
+    // Read and process all options (in an OPT RR) from a message.
+    // The `aOffset` points to beginning of record in `aMessage`.
+
+    otError  error = OT_ERROR_NONE;
+    uint16_t len;
+
+    IgnoreError(Dns::Name::ParseName(aMessage, aOffset));
+    aOffset += sizeof(Dns::OptRecord);
+
+    len = aOptRecord.GetLength();
+
+    while (len > 0)
+    {
+        Dns::LeaseOption leaseOption;
+        Dns::Option &    option = leaseOption;
+        uint16_t         size;
+
+        SuccessOrExit(error = aMessage.Read(aOffset, option));
+
+        VerifyOrExit(aOffset + option.GetSize() <= aMessage.GetLength(), error = OT_ERROR_PARSE);
+
+        if ((option.GetOptionCode() == Dns::Option::kUpdateLease) &&
+            (option.GetOptionLength() >= Dns::LeaseOption::kOptionLength))
+        {
+            SuccessOrExit(error = aMessage.Read(aOffset, leaseOption));
+
+            mAcceptedLeaseInterval = leaseOption.GetLeaseInterval();
+
+            if (mAcceptedLeaseInterval > kMaxLease)
+            {
+                mAcceptedLeaseInterval = kMaxLease;
+            }
+        }
+
+        size = static_cast<uint16_t>(option.GetSize());
+        aOffset += size;
+        len -= size;
+    }
+
+exit:
+    return error;
+}
+
+void Client::CheckState(void)
```

Comment:
nit: this name looks like the function has no side-effect. How about rename it to `UpdateState()` or `EvaluateState`?

Reply:
Good suggestion. I renamed it to `UpdateState()`.